### PR TITLE
SABR formula for normal volatilities

### PR DESCRIPTION
**Pull request by marc-henrard**
- Formula for implied normal/Bachelier volatility in the SABR model
- Add information on `SabrVolatilityFormula` to indicate if the volatility type is log-normal or normal.

### CODE REVIEW

**Comment by benhamilton-brown** [comment]
Added Stephen for the review the code and Christian to look at the SABR formula. I'll also add anything that I spot

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormula.java [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+  @Override
+  public ValueType getVolatilityType() {
+    return ValueType.NORMAL_VOLATILITY; // SABR implemented with normal/Bachelier implied volatility
+  }
+
+  //-------------------------------------------------------------------------
+
+  @Override
+  public double
+      volatility(
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormula.java:55]: Formatting

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormula.java [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+ * Formulas related to the SABR implied normal volatility function.
+ * <p>
+ * Only the "beta = 0" versions take strike/forward with negative values. The other formulas have a barrier at 0.
+ * <p>
+ * Reference: Hagan, P.; Kumar, D.; Lesniewski, A. & Woodward, D. "Managing smile risk", Wilmott Magazine, 2002, September, 84-108
+ * Note: Formula references are related to the preprint version, e.g. available at https://www.researchgate.net/publication/235622441_Managing_Smile_Risk
+ */
+@BeanDefinition(style = "light")
+public final class SabrHaganNormalVolatilityFormula
+    implements SabrVolatilityFormula, ImmutableBean, Serializable {
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormula.java:35]: Does this need to be a bean?

Comment (marc-henrard) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormula.java:35]: No it does not; there is no data in it. I copied the other version (`SabrHaganVolatilityFunctionProvider`) which is a Bean. Probably we should also remove the Bean feature on the other one.

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormulaTest.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    double[] time = {5.0, 10.0};
+    double[] alpha = {0.01, 0.02};
+    double[] rho = {-0.25, 0.50};
+    double[] nu = {0.25, 0.50};
+    for (int loopforward = 0; loopforward < forward.length; loopforward++) {
+      for (int loopstrikes = 0; loopstrikes < strikes.length; loopstrikes++) {
+        for (int looptime = 0; looptime < time.length; looptime++) {
+          for (int loopalpha = 0; loopalpha < alpha.length; loopalpha++) {
+            for (int looprho = 0; looprho < rho.length; looprho++) {
+              for (int loopnu = 0; loopnu < nu.length; loopnu++) {
```
Comment (ChristianOG) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormulaTest.java:119]: that's a whole lotta loops!

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormulaTest.java [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  @Test
```
Comment (ChristianOG) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormulaTest.java:289]: this is only for the beta <> 0.0 case, correct? negative forwards and strikes still work for beta = 0?

**Review by ChristianOG — COMMENTED** [review]
Just a comment about the tests for negative forwards and strikes

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormulaTest.java (hunk 4 above)
Comment (marc-henrard) [reply modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormulaTest.java:289]: Yes, for beta=0, negative forward and strikes are good. I'm adding some further tests with negative forwards and strikes in the "Beta0" case.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormula.java [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+  @Override
+  public ValueType getVolatilityType() {
+    return ValueType.NORMAL_VOLATILITY; // SABR implemented with normal/Bachelier implied volatility
+  }
+
+  //-------------------------------------------------------------------------
+
+  @Override
+  public double
+      volatility(
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormula.java:49]:
```suggestion
  public double volatility(
```

**Review by jodastephen — CHANGES_REQUESTED** [review]

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormula.java (hunk 5 above)
Comment (marc-henrard) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormula.java:49]: Sorry!

**Review by jodastephen — APPROVED** [review]

**Review by ChristianOG — APPROVED** [review]
Looks good to me.
